torch/utils/checkpoint.py

```diff
@@ -39,59 +39,68 @@ def backward(ctx, *args):
         return (None,) + tuple(inp.grad for inp in detached_inputs)
 
 
-def checkpoint(run_function, *args):
+def checkpoint(function, *args):
     r"""Checkpoint a model or part of the model
 
-    Checkpoint works by trading compute for memory. It can be applied on any
-    part of the model. In the forward pass, the model activations are not
-    stored. The forward pass save the inputs tuple and the run_function
-    parameter. In the backwards pass, the saved inputs and run_function is
-    retreived, and the forward pass is done on the model again (non-volatile
-    this time) since we need to get the activations values for calculating the
-    gradient and then the gradients are calculated.
+    Checkpointing works by trading compute for memory. Rather than storing all
+    intermediate activations of the entire computation graph for computing
+    backward, the checkpointed part does **not** save intermediate activations,
+    and instead computes them in backward pass of this particular part. It can
+    be applied on any part of a model.
+
+    Specifically, in the forward pass, :attr:`function` will run in
+    :func:`torch.no_grad` manner, i.e., not storing the intermediate
+    activations. Instead, the forward pass saves the inputs tuple and the
+    :attr:`function` parameter. In the backwards pass, the saved inputs and
+    :attr:`function` is retreived, and the forward pass is computed on
+    :attr:`function` again, now tracking the intermediate activations, and then
+    the gradients are calculated using these activation values.
 
     .. warning::
 
-        checkpointing doesn't work with torch.autograd.grad(), but only with
-        torch.autograd.backward()
+        Checkpointing doesn't work with :func:`torch.autograd.grad`, but only
+        with :func:`torch.autograd.backward`.
```

Comment:
Can we also add a warning that if `function` invocation in backward will do anything else than the one in forward (because e.g. a global has changed), then it won't be equivalent, and we won't be able to detect this?